### PR TITLE
remove special grad_mode tls handling

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63117
* __->__ #63116

This PR removes the special flag to disable grad mode tracking on the ThreadLocalState and replaces it with an explicit setter that users can use.
This allows to reduce complexity of ThreadLocalState.

Differential Revision: [D30388098](https://our.internmc.facebook.com/intern/diff/D30388098)

### CODE REVIEW

File: aten/src/ATen/ThreadLocalState.cpp [hunk 1]
```diff
@@ -8,32 +8,28 @@
...
 namespace at {
 
-ThreadLocalState::ThreadLocalState(bool keep_grad_mode)
+ThreadLocalState::ThreadLocalState()
     : dispatch_key_(c10::impl::tls_local_dispatch_key_set()),
       debug_info_(c10::ThreadLocalDebugInfo::current()),
       autograd_tls_(c10::AutogradTLS::get_mode()) {
   rf_tls_ = at::get_record_function_tls_();
-
-#if !defined(CAFFE2_IS_XPLAT_BUILD) && !defined(C10_MOBILE)
```
Comment (soulitzer) [inline aten/src/ATen/ThreadLocalState.cpp:17]: I guess this PR won't affect mobile/xplat builds, but just curious why we had this exception previously.

**Review by soulitzer — DISMISSED** [review]
Nice simplification!

**Comment by soulitzer** [comment]
Looks like updates need to be done in the distributed engine side as well

File: aten/src/ATen/ThreadLocalState.cpp (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/ThreadLocalState.cpp:17]: My guess is that the mobile build doesn't come with cuda so trying to set the GradMode would raise an error there.

**Comment by albanD** [comment]
For some reason ghstack didn't like my rebasing work and pushed a brand new stack (maybe I shouldn't have upgraded ghstack?)
In any case, the new version of this PR is here: https://github.com/pytorch/pytorch/pull/63318

Sorry for the noise.

**Review by soulitzer — APPROVED** [review]
LGTM

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388098).

**Pull request closed without merging**